development/workflow.rst

```diff
@@ -0,0 +1,305 @@
+To help you automatically authenticate when pushing to GitHub, we encourage you to follow the next step and enable a credential helper.
+
+.. _git-credential-helper:
+
+Setup a Git Credential Helper
+-----------------------------
+
+Rather than entering your GitHub username and password (or 2FA access token) every time you push, you can setup a Git credential helper to manage this for you.
+A credential helper is especially important for working with our :doc:`Git LFS-backed repositories </development/git_lfs>`.
+
+**Mac users** can use the secure OS X keychain:
+
+.. code-block:: bash
+
+   git config --global credential.helper osxkeychain  
+
+**Linux users** can use a credential *cache* to temporarily keep your credentials in memory.
+To have your credentials cached for 1 hour (3600 seconds):
+
+.. code-block:: bash
+
+   git config --global credential.helper 'cache --timeout=3600'
+
+**Linux users can alternatively** have their `credentials stored on disk <http://git-scm.com/docs/git-credential-store>`_ in a :file:`~/.git-credentials` file.
+Obviously, only do this for machines where you can ensure some level of security.
+
+.. code-block:: bash
+
+   git config credential.helper store
+
+Once a credential helper is enabled, the next time you ``git push``, you will add your credentials to the helper.
+
+Remember that if you have 2FA enabled, you will create and use a `personal access token <https://github.com/settings/tokens>`_ instead of your GitHub password.
+
+The DM Git LFS documentation has further information about :ref:`authenticating with our LFS storage backend <git-lfs-auth>`.
+
+.. _git-shell-setup:
+
+Tune your shell for Git
+-----------------------
+
+You can build an effective development environment and workflow by tuning your Git setup.
+Here are some ideas:
+
+1. `Add git status to your prompt <http://git-scm.com/book/en/v2/Git-in-Other-Environments-Git-in-Bash>`_.
+2. `Enable shell autocompletion <http://git-scm.com/book/en/v2/Git-in-Other-Environments-Git-in-Bash>`_
+3. `Craft aliases for common workflows <http://git-scm.com/book/en/v2/Git-Basics-Git-Aliases>`_.
+4. Use `hub <https://hub.github.com>`_ to interact with GitHub features from the command line.
+
+.. _git-commit-organization-best-practices:
+
+Appendix: Commit Organization Best Practices
+============================================
+
+Commits should represent discrete logical changes to the code
+-------------------------------------------------------------
+
+`OpenStack has an excellent discussion of commit best practices <https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes>`_; this is recommended reading for all DM developers.
+This section summarizes those recommendations.
+
+Commits on a ticket branch should be organized into discrete, self-contained, units of change.
+In general, we encourage you to err on the side of more granular commits; squashing a pull request into a single commit is an anti-pattern.
```

Comment:
The other side of the balancing act is that changes related to a single *feature* should be in the same commit, rather than spread over multiple commits.